scope_registry.go

```diff
@@ -152,23 +152,56 @@ func (r *scopeRegistry) Subscope(parent *scope, prefix string, tags map[string]s
 	// as the memory layout of []byte is a superset of string the below casting is safe and does not do any alloc
 	// However it cannot be used outside of the stack; a heap allocation is needed if that string needs to be stored
 	// in the map as a key
-	if s, ok := r.lockedLookup(subscopeBucket, *(*string)(unsafe.Pointer(&buf))); ok {
-		subscopeBucket.mu.RUnlock()
-		return s
+	var (
+		unsanitizedKey = *(*string)(unsafe.Pointer(&buf))
+		sanitizedKey   string
+	)
+
+	s, ok := r.lockedLookup(subscopeBucket, unsanitizedKey)
+	if ok {
+		// If this subscope isn't closed, return it. Otherwise, report it
+		// immediately and delete it so that a new (functional) scope can be
+		// returned instead.
+		if !s.closed.Load() {
+			subscopeBucket.mu.RUnlock()
+			return s
+		}
+
+		switch {
+		case parent.reporter != nil:
+			s.report(parent.reporter)
+		case parent.cachedReporter != nil:
+			s.cachedReport()
+		}
 	}
-	subscopeBucket.mu.RUnlock()
 
-	// heap allocating the buf as a string to keep the key in the subscopes map
-	preSanitizeKey := string(buf)
 	tags = parent.copyAndSanitizeMap(tags)
-	key := scopeRegistryKey(prefix, parent.tags, tags)
+	sanitizedKey = scopeRegistryKey(prefix, parent.tags, tags)
+
+	// If a scope was found above but we didn't return, we need to remove the
+	// scope from both keys.
+	if ok {
```

Comment:
The control flow here is a bit hard to grok for me; how hard/worse would it be if we move this into the above `if ok` check?

Reply:
It's essentially either this or an `if/else` with duplicated tag/key sanitization in both branches.

We only want to call `copyAndSanitizeMap` and `scopeRegistryKey` after we've done the lookup above and only if we're not returning a valid scope, so this felt preferable to duplicating code.

If it's helpful, I can expand on the comment a bit, but I think this is probably ok as-is.